src/common/snapdebugML.F90

```diff
@@ -0,0 +1,307 @@
+      real :: total_sys = 0.0
+#if defined(_OPENMP)
+      type(timer_womp_t) :: timer_womp
+      real :: total_womp = 0.0
+#endif
+      character(len=1024) :: prefix
+      contains
+        procedure :: start => prefixed_timer_start
+        procedure :: stop => prefixed_timer_stop
+        procedure :: stop_and_log => prefixed_timer_stop_and_log
+        procedure :: print_accumulated => prefixed_timer_print_accumulated
+    end type
+
+    interface prefixed_accumulating_timer
+      procedure :: prefixed_accumulating_timer_new
+    end interface
+
+    contains
+
+    function timer_cpu_new()
+      type(timer_cpu_t) :: timer_cpu_new
+
+      call cpu_time(timer_cpu_new%innertime)
+      if (timer_cpu_new%innertime == -1.0) then
+        write(error_unit,*) "No cpu time available, reported times are not reliable"
+      endif
+    end function
+
+    !> Returns the time in seconds since initializing this clock
+    function timer_cpu_now(this)
+      class(timer_cpu_t), intent(in) :: this
+      real :: timer_cpu_now
+      call cpu_time(timer_cpu_now)
+
+      timer_cpu_now = timer_cpu_now - this%innertime
+    end function
+
+    !> Gets current time since start and
+    !> outputs elapsed time into the log
+    subroutine timer_cpu_log(this, prefix, unit)
+      class(timer_cpu_t), intent(in) :: this
+      character(len=*), intent(in), optional :: prefix
+      integer, intent(in), optional :: unit
+      real :: now
+
+      now = this%now()
+      call print_time(now, prefix, unit)
+    end subroutine
+
+    function timer_sys_new()
+      type(timer_sys_t) :: timer_sys_new
+
+      call system_clock(count=timer_sys_new%count, count_rate=timer_sys_new%countrate)
+    end function
+
+    function timer_sys_now(this)
+      class(timer_sys_t), intent(in) :: this
+      real :: timer_sys_now
+      integer :: now_count
+
+      call system_clock(count=now_count)
+      now_count = now_count - this%count
+      timer_sys_now = real(now_count) / real(this%countrate)
+    end function
+
+    subroutine timer_sys_log(this, prefix, unit)
+      class(timer_sys_t), intent(in) :: this
+      character(len=*), intent(in), optional :: prefix
+      integer, intent(in), optional :: unit
+      real :: now
+
+      now = this%now()
+      call print_time(now, prefix, unit)
+    end subroutine
+
+#if defined(_OPENMP)
+    function timer_womp_new()
+      type(timer_womp_t) :: timer_womp_new
+
+      timer_womp_new%inner = omp_get_wtime()
+    end function
+
+    function timer_womp_now(this)
+      class(timer_womp_t), intent(in) :: this
+      real :: timer_womp_now
+      double precision :: t
+
+      t = omp_get_wtime()
+
+      timer_womp_now = t - this%inner
+    end function
+#endif
+
+    function to_str(t) result(str)
+      real, intent(in) :: t
+      character(len=64) :: str
+
+      integer :: h, m, s, ms
+      call real_to_hms_ms(t, h, m, s, ms)
+
+      write(str,"(I3, a, I2, a, I2, a, I4)") h, ":", m, ":", s, ".", ms
```

Comment:
might be more elegant as
`write(str,"(I3, ':', I0.2, ':', I0.2, '.', I04)") h, m, s, ms`

Reply:
`I0.4`